# Error analysis - example 8.1

The values of $y=x^{1/3}$ and $z=x^{30}$ are calculated for $x=0.92$ which is correctly rounded. One gets $y=0.9726$ and $z=0.08197$. How many correct digits have the results? With how many correct digits should $x$ be given in order to get all digits in $z$ correctly.

In [1]:
x=0.92
y=0.9726
z=0.08197

## Number of correct digits in results

The absolute and relative error limits for $x$ are given by

In [2]:
E_x = 0.005
R_x = abs(E_x/x,)
print "E_x=%.4f, R_x=%.4f" % (E_x, R_x)

E_x=0.0050, R_x=0.0054


The relative error's limit for an exponentiation $z=x^n$ is given by $R_z=|n|R_x$.

<hr>
We get for $y=x^{1/3}$

In [3]:
R_y=abs(1./3)*R_x; E_y=y*R_y;
print 'R_y=%.4f    E_y=%.4f' % (R_y, E_y)

R_y=0.0018    E_y=0.0018


The result becomes $y=0.92^{1/3}=0.9726 \pm 0.0018$, i.e. only two digits of the result are guaranteed to be correct.

<hr>
We get for $z=x^{30}$

In [4]:
R_z=abs(30)*R_x; E_z=z*R_z;
print 'R_z=%.4f    E_z=%.5f' % (R_z, E_z)

R_z=0.1630    E_z=0.01336


The result becomes get $z=0.92^{30}=0.08197 \pm 0.01336$, i.e. no digit of the result is guaranteed to be correct.

## Required precision of input

We calculate now the required number of correct digits for x in order to get 6 correct digits in order to get 4 correct digits resp. 5 correct decimals in the result $z$.

The limit of the relative error for 4 signifikant digits in z is therefore

In [5]:
R_4 = 5E-4

This leads to a maximum of the relativ error in the argument $x$

In [6]:
R_x = R_4/30; print "R_x = %.1e" % R_x

R_x = 1.7e-05


We find $R_x < 5 \cdot 10^{-5}$, i.e. 5 correct digits of the argument x would lead to 4 correct digits in this calculation.

<hr>
However, we know that in general $0.5 \cdot 10^{-4} \leq R_x \leq 5 \cdot 10^{-4}$. To be sure to get 4 correct digits in the result for any possible argument we need therefore to calculate the smallest required limit of the relative error.

In [2]:
R_4 = 0.5E-4
R_x = R_4/30; print "R_x = %.1e" % R_x

R_x = 1.7e-06


We find $R_x < 5 \cdot 10^{-6}$, i.e. 6 correct digits of the argument x would
lead to 4 correct digits in the result for any argument $x$.